### PR TITLE
Refactor randomness tests

### PR DESCRIPTION
**Pull request by samdow**
Refactors the randomness tests based on their input type + OOP vs in-place. So this makes a new test for:
- OOP, takes in shape
- OOP, takes in one tensor
- OOP, takes in two tensors (also tests if either tensor is batched as well as both)
- in place, takes in one tensor
- in place, takes in two tensors (this is only bernoulli, also tests if either tensor is batched as well as both)

It also leaves the randperm + dropout tests because those are special cased

### CODE REVIEW

File: test/test_vmap.py [hunk 1]
```diff
@@ -3466,57 +3466,81 @@ def f(x, y):
...
+            vmap(fn, in_dims=in_dims, randomness="different")(*args)
+
+    def _assert_throws_in_same_mode_batched(self, fn, args, in_dims=0):
+        with self.assertRaisesRegex(RuntimeError,
+                                    r"Vmap does not currently support same randomness with a batched tensor input"):
+            vmap(fn, in_dims=in_dims, randomness="same")(*args)
+
+    def _in_dims(self, *batched):
+        batched = batched + (True,)  # for the always batched dummy argument
+        return tuple(0 if is_batched else None for is_batched in batched)
```
Comment (zou3519) [inline test/test_vmap.py:3505]: Doesn't need to happen in this PR, but we should test a case where, if the tensor already has a dimension being vmapped over, that dimension is not at the front of the tensor (so, not zero)

File: test/test_vmap.py [hunk 2]
```diff
@@ -3466,57 +3466,81 @@ def f(x, y):
...
+
     @parametrize('randomness', ['same', 'different', 'error'])
     @parametrize('use_generator', [True, False])
-    def test_random_behavior(self, device, randomness, use_generator):
+    def test_shape_OOP_ops(self, device, randomness, use_generator):
+        # ops that take in the final shape and are out of place
```
Comment (zou3519) [inline test/test_vmap.py:3510]: nit: oop is generally the abbreviation for "object-oriented-programming". By "shape oop" you probably mean "factory" -- a factory function is a function that accepts no Tensor arguments and returns a Tensor. Maybe let's call it `test_factory_ops` ?

File: test/test_vmap.py [hunk 3]
```diff
@@ -3746,49 +3754,206 @@ def test_like_functions(self, device, randomness, batched_input):
 
     @parametrize('use_generator', [True, False])
     @parametrize('randomness', ['error', 'same', 'different'])
-    def test_random_inplace_not_batched(self, device, use_generator, randomness):
-        # tests that when in place random ops are being called during vmap but not with a batched tensor,
-        # it behaves like a normal random_ when using same random and errors during different random
-        supported_ops = [
+    @parametrize('batched_input', [True, False])
+    def test_random_inplace_single(self, device, use_generator, randomness, batched_input):
```
Comment (zou3519) [inline test/test_vmap.py:3758]: nit: test_random_unary_inplace: "unary" is shorthand for "single tensor input"

File: test/test_vmap.py [hunk 4]
```diff
@@ -3746,49 +3754,206 @@ def test_like_functions(self, device, randomness, batched_input):
...
+        orig_state = generator.get_state()
+        kwargs = {'generator': generator} if use_generator else {}
+        p = torch.rand(14)
+        ops = [
             lambda t, _: t.random_(**kwargs),
             lambda t, _: t.random_(100, **kwargs),
             lambda t, _: t.random_(-5, 100, **kwargs),
             lambda t, _: t.normal_(**kwargs),
-            lambda t, _: t.bernoulli_(torch.tensor([0.3, 0.4, 0.5, 0.6]), **kwargs),
+            lambda t, _: t.bernoulli_(p, **kwargs),
```
Comment (zou3519) [inline test/test_vmap.py:3768]: this is not really a single Tensor input, right? Because t is a Tensor and p is also a Tensor. Should this be here?

File: test/test_vmap.py [hunk 5]
```diff
@@ -3746,49 +3754,206 @@ def test_like_functions(self, device, randomness, batched_input):
...
+            expected = op(input_expected, probability, always_batched)
+            self._assert_all_slices_equal(vmap_result)
+            for i in range(B0):
+                self.assertEqual(vmap_result[i], expected)
+
+    @parametrize('use_generator', [True, False])
+    @parametrize('randomness', ['error', 'same', 'different'])
+    @parametrize('batched_input', [True, False])
+    @parametrize('batched_other', [True, False])
+    def test_random_out_of_place_two_tensors(self, device, use_generator, randomness, batched_input, batched_other):
```
Comment (zou3519) [inline test/test_vmap.py:3861]: nit: test_binary_outplace -- "binary" is the official term for a function that accepts two tensor arguments. But also feel free to ignore, it's pretty clear from this test name what is going on

**Review by zou3519 — APPROVED** [review]
The tests look great after the cleanup!

One minor nit: on vocabulary: "unary" can be shorthand for "function that accepts single Tensor argument", "binary" can be shorthand for "function that accepts two Tensor arguments", "factory" can be shorthand for "function that accepts no Tensor arguments but returns a Tensor". Your tests names are already pretty descriptive so feel free to ignore.

Next, it would be nice if we tested a tensor where the dimension being vmapped over isn't the front (right now it's hardcoded in as (B, 3, 3, 14, 14). But this doesn't need to happen in this PR (it's better as a follow-up)

File: test/test_vmap.py (hunk 4 above)
Comment (samdow) [reply test/test_vmap.py:3768]: Good catch! Not certain how that happened but removed this and added kwargs to the bernoulli_in_place test that does take in a tensor for probability

File: test/test_vmap.py (hunk 5 above)
Comment (samdow) [reply test/test_vmap.py:3861]: Fixed

File: test/test_vmap.py (hunk 3 above)
Comment (samdow) [reply test/test_vmap.py:3758]: Fixed

File: test/test_vmap.py (hunk 2 above)
Comment (samdow) [reply test/test_vmap.py:3510]: Fixed

File: test/test_vmap.py (hunk 1 above)
Comment (samdow) [reply test/test_vmap.py:3505]: Sounds good! I'll get another PR for that. There will be another follow up since this already raised a bug with multinomial so good to check these things

**Review by zou3519 — APPROVED** [review]
